### ISSUE TITLE
repeated panics when using ServiceExt::oneshot

### ISSUE THREAD

**Issue by hdevalence** [tower-rs/tower#469]
Using `ServiceExt::oneshot`  (tower 0.3.1) I get repeated panics with the message "We immediately transition to ::Called".

It looks like this comes from the service impl:
```rust
impl<S, Req> Future for Oneshot<S, Req>
where
    S: Service<Req>,
{
    type Output = Result<S::Response, S::Error>;

    #[project]
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let mut this = self.project();
        loop {
            #[project]
            match this.state.as_mut().project() {
                State::NotReady(nr) => {
                    let (mut svc, req) = nr.take().expect("We immediately transition to ::Called");
                    let _ = ready!(svc.poll_ready(cx))?;
                    this.state.set(State::Called(svc.call(req)));
                }
                State::Called(fut) => {
                    let res = ready!(fut.poll(cx))?;
                    this.state.set(State::Done);
                    return Poll::Ready(Ok(res));
                }
                State::Done => panic!("polled after complete"),
            }
        }
    }
}
```

I'm not sure I follow the reasoning here: `svc` is taken from the `nr: Option<_>`, but if `svc.poll_ready` returns `Poll::Pending`, `poll` will return early (via `ready`) without putting `svc` back.  So I don't see how it immediately transitions to `::Called`.

**Comment by hawkw** [comment]
I believe this is the same bug that was fixed in #447. Maybe that was not backported to 0.3?

**Comment by LucioFranco** [comment]
@hdevalence I think you all updated to 0.4, is this still an issue or can we close this?

**Comment by hawkw** [comment]
This is fixed in 0.4.
